platform/darwin/src/NSComparisonPredicate+MGLAdditions.mm

```diff
@@ -361,9 +142,52 @@ - (id)mgl_jsonExpressionObject {
                         format:@"NSPredicateOperatorType:%lu is not supported.", (unsigned long)self.predicateOperatorType];
     }
     if (op) {
-        return @[op, self.leftExpression.mgl_jsonExpressionObject, self.rightExpression.mgl_jsonExpressionObject];
+        id leftJSONExpressionObject = self.leftExpression.mgl_jsonExpressionObject;
+        id rightJSONExpressionObject = self.rightExpression.mgl_jsonExpressionObject;
+        
+        switch (self.predicateOperatorType) {
+            case NSLessThanPredicateOperatorType:
+            case NSLessThanOrEqualToPredicateOperatorType:
+            case NSGreaterThanPredicateOperatorType:
+            case NSGreaterThanOrEqualToPredicateOperatorType: {
+                leftJSONExpressionObject = [self mgl_typedJSONWithExpressionObject:self.leftExpression fallbackExpression:self.rightExpression];
+                rightJSONExpressionObject = [self mgl_typedJSONWithExpressionObject:self.rightExpression fallbackExpression:self.leftExpression];
+                break;
+            }
+ 
+            default:
+                break;
+        }
+        
+        return @[op, leftJSONExpressionObject, rightJSONExpressionObject];
     }
     return nil;
 }
 
+/**
+ Infers the expressionObject type if it can not then tries to infer the type using fallbackExpression.
+ ExpressionFilters of type >, >=, <, <= requires that each compared value provides its type.
+ */
+- (id)mgl_typedJSONWithExpressionObject:(NSExpression *)expressionObject fallbackExpression:(NSExpression *)fallbackExpression {
+    NSExpression *expression = expressionObject.expressionType == NSConstantValueExpressionType ? expressionObject : fallbackExpression;
```

Comment:
Replace `expressionObject` with `self` and `fallbackExpression` with `otherExpression`.